### ISSUE TITLE
Value precedence not working with child groups

### ISSUE THREAD

**Issue by s2hc-johan** [ansible/ansible#6489]
##### Issue Type:

Bug
##### Ansible Version:

Have encountered it in head, 1.5.2 and 1.5.3, according to discussions on irc it has been around for quite some time.
##### Environment:

Deploy os is Os X mavericks servers deployed to is debian 7.
##### Summary:

When using child groups to be able to override som variables, the variable from the parent group is the one that gets used. 

All tough this bug is not explicitly described in the pull request https://github.com/ansible/ansible/pull/6379 it did solve the problem. Tested the PR when merging with 1.5.3 branch.
##### Steps To Reproduce:

small example snippet (I can add full files if needed).

Inventory file:

```
[devproxyservers]
10.0.40.83
10.0.40.84

[proxyservers:children]
devproxyservers
```

proxyservers group vars

```
cert_path: "{{ base_dir }}/certs/{{ cert_name }}"
key_path: "{{ base_dir }}/certs/{{ key_name }}"
```

devproxyservers group vars

```
cert_path: "{{ base_dir }}/certs/fake_cert.crt"
key_path: "{{ base_dir }}/certs/fake_key.key"
```
##### Expected Results:

according to docs http://docs.ansible.com/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable the cild group vars should be used.
##### Actual Results:

The variables of proxyservers group is the variable that gets used

**Comment by srgvg** [comment]
The full and exact inventory where you experienced the problem will be needed to reproduce the issue; with just the snippet chances are high the bug doesn't show.

**Comment by s2hc-johan** [comment]
The full inventory file looks like:

```
[all:vars]
env=dev

[devaserviceservers]
10.0.40.85

[devbserviceservers]
10.0.40.86

[serviceservers:children]
devaserviceservers
devbserviceservers

[webservers]
10.0.40.83
10.0.40.84

[appservers:children]
webservers
serviceservers

[dbservers]
10.0.40.82

[devproxyservers]
10.0.40.83
10.0.40.84

[proxyservers:children]
devproxyservers

[mqservers]
10.0.40.81

[monservers:children]
dbservers
proxyservers
mqservers
webservers
serviceservers

[development:children]
dbservers
proxyservers
mqservers
webservers
serviceservers
```

**Comment by srgvg** [comment]
FYI This is the problem solved by 2f2a22b87e33f39286d0b2daa5fd50c60fd4788b

**Comment by mpdehaan** [comment]
Thanks @sergevanginderachter -- since we have this ticket open I'm going to close this one as a duplicate and mark your ticket as a priority one so we are sure to look at it quicker.

Thanks!

**Comment by mpdehaan** [comment]
I see this is applied to devel, so we should be good to go!

**Comment by srgvg** [comment]
PR or single commit wasn't applied yet, not sure what you meant.

**Comment by alanfairless** [comment]
Ran into this earlier. I can confirm the patch above by @sergevanginderachter resolves it, and I don't see that commit as part of `devel`.  Could you try reapplying?

**Comment by mpdehaan** [comment]
Ok, all being said, the pull request is still open so we only need the one ticket - the PR will suffice.

**Comment by basetta** [comment]
@mpdehaan i have the same issue in 1.8.4. I did not find the commit 2f2a22b87e33f39286d0b2daa5fd50c60fd4788b in the following branch: devel, release1.8.4. Can you please point me out where is the pull-request ?

**Comment by srgvg** [comment]
That patch got rebased at some point, I think. Look for cc8efb4 Was merged in 1.7

If you experience thos on 1.8.4, this might be another issue. Perhaps post your full details in a new issue, or post it to the mailing list first.

**Comment by sebastiaanluca** [comment]
Has this been fixed? Having the exact same issue.

Inventory:

```
[web]
xxx.xxx.xxx.xxx

[database]
xxx.xxx.xxx.xxx

[local:children]
web
database
```

`/group_vars/web/default.yaml`:

```
my_variable: "123"
```

`/group_vars/local/default.yaml`:

```
my_variable: "abc"
```

Result of `my_variable` is `123` when it should be `abc`, right?

Explicitly defining the server in a block works as expected, yet I want to add groups, not individual servers.

```
[local]
xxx.xxx.xxx.xxx
```

---

Hang on… I think I misread the example in the initial post. So my example actually works as intended while I want the inverse (bugged flow) to happen?
